storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,811 @@
+        val fsmState = fsmStateAndIteration._1
+        val iteration = fsmStateAndIteration._2
+
+        if (iteration < multiplier) {
+          val transition = transitions.get(fsmState)
+          if (transition.nonEmpty) {
+            val subState = transition.get.get(symbol)
+            if (subState.nonEmpty) {
+              next += subState.get -> iteration
+              //final of self? merge with initial on next iteration
+              if (finalStates.contains(subState.get)) {
+                next += initialState -> (iteration + 1)
+              }
+            }
+          }
+        }
+      })
+
+      if (next.nonEmpty) {
+        Some(next)
+      } else {
+        None
+      }
+    }
+
+    //If the initial state is final then multiplying doesn't alter that
+    def isFinal(crawlState: HashSet[(Int, Int)]): Boolean =
+      crawlState.exists(fsmStateAndIteration =>
+        (fsmStateAndIteration._1 == initialState)
+          && (finalStates
+            .contains(initialState) || (fsmStateAndIteration._2 == multiplier)))
+
+    Fsm.crawl(alphabet, initial, isFinal, follow).Reduced
+  }
+
+  /**
+    * If the present FSM accepts X, returns an FSM accepting X* (i.e. 0 ormore Xes).
+    * This is NOT as simple as naively connecting the final states back to the initial state:
+    * see (b*ab)* for example.
+    */
+  def Star: Fsm = {
+    def follow(subStates: HashSet[Int], symbol: Char): Option[HashSet[Int]] = {
+      val next = HashSet[Int]()
+
+      for (subState <- subStates) {
+        val transition = transitions.get(subState)
+        if (transition.nonEmpty) {
+          val nextState = transition.get.get(symbol)
+          if (nextState.nonEmpty) {
+            next += nextState.get
+          }
+        }
+        //If one of our substates is final, then we can also consider
+        //transitions from the initial state of the original FSM.
+        if (finalStates.contains(subState)) {
+          val transition = transitions.get(initialState)
+          if (transition.isDefined) {
+            val nextState = transition.get.get(symbol)
+            if (nextState.nonEmpty) {
+              next += nextState.get
+            }
+          }
+        }
+      }
+
+      if (next.nonEmpty) {
+        Some(next)
+      } else {
+        None
+      }
+    }
+
+    def isFinal(subStates: HashSet[Int]): Boolean = subStates.exists(subState => finalStates.contains(subState))
+
+    Fsm.crawl(alphabet, HashSet(initialState), isFinal, follow) | Fsm.EpsilonFsm(alphabet)
+  }
+
+  /**
+    * Consider the FSM as a set of strings and return the cardinality of thatset
+    * - Some[Int], or None if there are infinitely many
+    */
+  def Cardinality: Option[Int] = {
+    val stateToCount = HashMap[Int, Option[Int]]()
+
+    def getNumStrings(state: Int): Option[Int] =
+      if (IsLive(state)) {
+        val stateCount = stateToCount.get(state)
+
+        if (stateCount.nonEmpty) {
+          if (stateCount.get.isEmpty) {
```

Comment:
Can you use `getOrElse` or something like `flatMap` here?

Reply:
getOrElse looks better option in this case, updated, also removed mutable n variable